Jbzdak/deploy databases public

jbzdak opened the pull request:
# Testing instructions

Please follow readmes in the secrets repository: https://github.com/open-craft/deployment-secrets-databases/blob/jbzdak/secrets/README.md
# Author's concerns
- [x] ~We will have two secret repositorues (or deployment repositories) one for databaes and one for servers. The problem is that in most cases they should be ran both at the same time, before you deploy a server you usually need to deploy the database as well.~
- [ ] Requirements file [looks atrociously bad](https://github.com/open-craft/deployment-deploy-databases/blob/jbzdak/mysql-deployment/requirements.yml). It will be much better if we register all our roles to a `ansible-galaxy`, then dependencies of our roles will be deployed automatically (at the very least we'll get rid of hashes there). Ansible role dependencies support versioning, albeit using different syntax than `requirements.yml`: 

```
       dependencies:
         - { role: 'git+https://bitbucket.org/ursuad/sample_role_launchpad-ppa-webupd8,v1.0.0'}
```
# Post approval actions
- [x] Squash commits
- [x] Replace all passwords
- [x] Replace all tarsnap keys, 
- [ ] and nuke old backups
- [x] Replace all snitches
- [x] Pin all requirements to hashes (instead of branches)
- [x] Change SMTP
- [ ] Change recipient e-mail 
- [x] Add tarsnap role as a dependency to `common-server` so we get failed provisioning if no backups is configured.

antoviaque commented on README.md:
```diff
@@ -1 +1,48 @@
+
+1. Create `private-extra-vars.yaml`, you'll put all the generated variables there
+1. Generate root password, put it in the `private.yaml` for your host, under `MYSQL_ROOT_PASSWORD`. 
+2. Create a key, but store it somewhere save (for example keepassx database), this key shouldn't end on the mysql server.   
```
"safe"

antoviaque replied:
This is still not addressed either!

antoviaque commented on hosts:
```diff
@@ -0,0 +1,2 @@
+[mysql]
+mysql.opencraft.com ansible_host=149.202.171.102
```
The hosts files could be on the private repo, though. It's not too important since this info will also be obtainable through DNS.

antoviaque commented on group_vars/mysql/public.yml:
```diff
@@ -0,0 +1,30 @@
+  - mount: /
+    treshold_mb: 5000
+  - mount: /var/lib/mysql
+    treshold_mb: 5000
```
We talked about using % here no?

jbzdak replied:
Damn :( See [this one](https://github.com/open-craft/ansible-sanity-checker/pull/1/files/8c7a536bc88867fd75f7081b7a806ae811a86fd3#r60913763). I forgot to update variables here (which means that this role is using default of 20%)

antoviaque commented:
@jbzdak From testing:

* Stock version of Ubuntu
* `mkfs.ext4 -j ...` for journalizing it
* The submodules version seem wrong, at least on the databases secret repo:

```
antoviaque@antoviaque-laptop:~/prog/edx/devops/deployment-secrets-services$ git submodule update
Cloning into 'deploy'...
remote: Counting objects: 22, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 22 (delta 0), reused 0 (delta 0), pack-reused 19
Receiving objects: 100% (22/22), 6.60 KiB | 0 bytes/s, done.
Checking connectivity... done.
fatal: reference is not a tree: 80764baf747e59df1ae558d17f8024a0728f2681
Unable to checkout '80764baf747e59df1ae558d17f8024a0728f2681' in submodule path 'deploy'
```

* `pip install -r requirements` fails on my Ubuntu - you might be missing some dependencies as debian packages? Cf https://gist.github.com/antoviaque/60b4f497433097a4df0e6c91a3e0aca6

jbzdak commented:
@antoviaque Just to clarify: uploading of ubuntu image should be a documented (in secrets repository) manual step which I need to take in Dev, Dalite, and Databases projects? 

Submodules issues are solved.

antoviaque commented:
@jbzdak 

> Just to clarify: uploading of ubuntu image should be a documented (in secrets repository) manual step which I need to take in Dev, Dalite, and Databases projects? 

Yup, something to document in the testing steps (it can be public, right? why in the secrets repo?), and to do when deploying the actual setup, to ensure we use the right version.

antoviaque commented on group_vars/all/public.yml:
```diff
@@ -0,0 +1,5 @@
+tarsnap_cache: /var/tarsnap/cache
+tarsnap_cron_hour: "*"
+
+FORWARD_MAIL_RELAY_MAIL_TO: 'jacek@opencraft.com'
```
Can you change the (public) default to a non-functioning addres, like `ops@example.com`, so that we don't get mail if others reuse our roles? Then override it in the private repo.

jbzdak replied:
Addressed in: d4db75aa3d3c37e4565b65605a62786abc8d943e

antoviaque commented:
@jbzdak 

* I don't see the update to the test instructions for the base Ubuntu image to use? For now I'm using the ones provided by OVH, but could you update the instructions (including the upload of the image via glance) & make sure you're using the right base image for the dalite migration dry run task?

> In region B create a VM for dalite. This one really should be bigger than inital 20GB, but I
didn't manage to get it working. 

I don't think you can currently change the base size - however this would be added as a volume, like on the mysql instance. I would at least put the contents of logs-downloads on a volume, so we can rebuild a VM without clearing the history of logs.

* The instructions above in "Initial setup" are outdated (paths are wrong with the public/private repos split), and don't contain the full command for the ansible-galaxy setup.

* There is a warning while installing the ansible dependencies, apparently

```
antoviaque@antoviaque-laptop:~/prog/edx/devops/deployment-secrets-databases$ ./venv/bin/ansible-galaxy install -r deploy/requirements.yml -f
- extracting geerlingguy.mysql to deploy/roles/geerlingguy.mysql
- geerlingguy.mysql was installed successfully
- extracting geerlingguy.security to deploy/roles/geerlingguy.security
- geerlingguy.security was installed successfully
- extracting kamaln7.swapfile to deploy/roles/kamaln7.swapfile
- kamaln7.swapfile was installed successfully
- extracting pmbauer.tarsnap to deploy/roles/pmbauer.tarsnap
- pmbauer.tarsnap was installed successfully
- extracting backup-to-tarsnap to deploy/roles/backup-to-tarsnap
- backup-to-tarsnap was installed successfully
[DEPRECATION WARNING]: The comma separated role spec format, use the yaml/explicit format instead.. This feature will be removed in a future release. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
- dependency pmbauer.tarsnap already pending installation.
- extracting common-server to deploy/roles/common-server
- common-server was installed successfully
- dependency geerlingguy.security already pending installation.
- dependency kamaln7.swapfile already pending installation.
- extracting configure-apt to deploy/roles/configure-apt
- configure-apt was installed successfully
- extracting sanity-checker to deploy/roles/sanity-checker
- sanity-checker was installed successfully
- extracting forward-server-mail to deploy/roles/forward-server-mail
- forward-server-mail was installed successfully
- extracting mysql to deploy/roles/mysql
- mysql was installed successfully
```

* Where is the internal part of the documentation? https://github.com/open-craft/deployment-deploy-databases/pull/2/files#diff-04c6e90faac2675aa89e2176d2eec7d8R28 is generic and you closed the internal doc PR saying you would put up a new one with what is specific to us, but I haven't seen it. For now I'm using the test instructions from the PR description above, but we need something more permanent for this. Also worth including helper commands such as `ansible-vault edit private-extra-vars.yaml --vault-password-file .vault-pass` to edit the private env files.

* When I run the playbook, I am getting the following error:

```
TASK [sanity-checker : copy sanity check config] *******************************
fatal: [mysql.opencraft.hosting]: FAILED! => {"failed": true, "msg": "'MYSQL_SANITY_CHECK_SNITCH' is undefined"}
```

`private-extra-vars.yaml` seem to have that variable defined (I'm using the private repo yml file). I wondered if it was that the command-line provided in the current PR description was out of date, so I have tried with both:

```
./venv/bin/ansible-playbook deploy/deploy-all.yml --vault-password-file .vault-pass -u admin
```

and 

```
./venv/bin/ansible-playbook deploy/deploy-all.yml --vault-password-file .vault-pass -u admin --extra-vars private-extra-vars.yaml
```

* There is a warning during the playbook execution:

```
TASK [configure-apt : Check if packages need to be autoremoved] ****************
ok: [mysql.opencraft.hosting]
 [WARNING]: Consider using apt-get module rather than running apt-get
```

jbzdak commented:
@antoviaque 

> I don't think you can currently change the base size - however this would be added as a volume, like on the mysql instance. I would at least put the contents of logs-downloads on a volume, so we can rebuild a VM without clearing the history of logs.

This is a good idea, howerver right now dalite is deployed on 40GB, and we should get notified when disk space exceeds 32GB. Adding a volume for user download logs is not overly complex --- however these "not so complex" tasks have been piling up from the start of the issue, and I'd rather have this done as a separate follow up fix. We can always restore the logs from backups. 

> There is a warning while installing the ansible dependencies, apparently

This warning is triggered by some error in ansible checking logic: https://github.com/ansible/ansible/pull/14612 , it was harmless. nevertheless it is now fixed 

> There is a warning during the playbook execution:

This warning is caused by calling `apt-get autoremove` which was in plebia playbooks. At the time of writing this role there were no way to avoid this warning. Ansible 2.1 has this feature, but it was released 20 hours ago. I'd rather not update Ansible just now. This would be good canditate for a follow-up task. 

I have moved deployment instructions to two readme files, please use them, nevertheless following combination of commands worked for me on empty directory 

* `git clone git@github.com:open-craft/deployment-secrets-databases.git`
* `cd deployment-secrets-databases`
* `git checkout jbzdak/secrets`
* Activate virtualenv with ansible
* Copy vault password
* `ansible-galaxy install -r deploy/requirements.yml -f`
* `ansible-playbook deploy/deploy-all.yml --vault-password-file .vault-pass -u ubuntu --extra-vars @private-extra-vars.yaml`

antoviaque commented:
@jbzdak They put a deprecation warning for a feature which doesn't have a replacement yet? Ie there is no way to cleanup that warning? The ansible maintainers are drunk.

jbzdak commented:
@antoviaque They put deprecation warning on every usage of ``apt-get`` in shell and/or command modules (among others)

antoviaque commented on README.md:
```diff
@@ -1 +1,50 @@
+Create a OpenStack environment
+------------------------------
+
+1. Create 10GB volume for root filesystem, create it using image "Ubuntu 14.04". 
```
Missing the use of a vanilla image.

jbzdak replied:
Addressed in 2b072e2b8cefd9cb70660bab73bccddc978d52a5

antoviaque commented on README.md:
```diff
@@ -1 +1,50 @@
+    2. Outgoing traffic should be allowed. 
+    3. When you provision servers using the database allow them using their IP.
+3. After volume for root fs is ready create a VM for the mysql:
+    1. It should use root volume as filesystem (`Boot Source` -> `Volume` -> `Your root volume`)
```
@jbzdak You did this on the MySQL instance too? Wasn't it properly handled before, assigning just the right folder to store the MySQL DB data, but not the rest of the instance?

jbzdak replied:
@antoviaque No. Mysql instance works mounted on two volumes, first in on `/dev/vda` and translates to `/`, second is on `/dev/vdb` and is mounted on `/var/lib/mysql`.

jbzdak replied:
@antoviaque  addressed in 2b072e2b8cefd9cb70660bab73bccddc978d52a5

antoviaque commented on README.md:
```diff
@@ -1 +1,50 @@
+Perform deployment
+------------------
+
+       ansible-galaxy install -r requirements.yml -f && ansible-playbook deploy-all.yml  -u admin --extra-vars @private-extra-vars.yaml
```
On vanilla VMs, the user is `ubuntu`, not `admin`.

Also, the naming of the private file is inconsistent with the private doc - it's `private-extra-vars.yml`

jbzdak replied:
Addressed in: 4b12553bd77981e00214783b00caa4fd043d881f

antoviaque commented:
:+1:

Commit:
Initial version of database provisioning

Squashed commit of the following:

commit 058598f55d6d91bc9f379f8dce0a3153bdbabaad
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Fri Apr 29 21:23:46 2016 +0200

    Fix for emails sent from initial provisioning

commit 4b12553bd77981e00214783b00caa4fd043d881f
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Fri Apr 29 15:15:05 2016 +0200

    Updated readme

commit 2b072e2b8cefd9cb70660bab73bccddc978d52a5
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Fri Apr 29 15:03:37 2016 +0200

    Updated readme

commit 03f0727d7d7576dd5ce8a827748daffb247510a6
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Wed Apr 27 19:05:03 2016 +0200

    Added license file

commit d4db75aa3d3c37e4565b65605a62786abc8d943e
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Wed Apr 27 17:16:33 2016 +0200

    Added nonsencial value for: FORWARD_MAIL_RELAY_MAIL_TO.

commit 084df43369e47895c185a24e7ccea3ab80e602b6
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Wed Apr 27 17:13:00 2016 +0200

    Replaced `admin` user with generic term: administrative user.

commit 0aa03fd1f04bf37f0e8272a249149313fc5089c3
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Wed Apr 27 16:25:00 2016 +0200

    Updated readme

commit 004a109db7a43e14b23ceb50258295e321ae0a50
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Tue Apr 26 16:48:29 2016 +0200

    Fixed typo

commit b103e5012c1c81544b18e0d3ce8bd114a758e29e
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Tue Apr 26 16:35:14 2016 +0200

    Added snitches, and renamed instance to opencraft.hosting

commit 3d3768cfc259bb0519abe07a4fdd1a644a674c81
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 18:49:34 2016 +0200

    Fixed instructions

commit 7cd4efdf1e5e061e3394f08d52c35757362d53e3
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 18:32:15 2016 +0200

    Removed unneccessary requirement

commit 87067429cfaeb62dd4edfd8302f3f65ebb1c6286
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 15:50:51 2016 +0200

    Removed old check variables

commit 584e48ed276cf273a9feb9e2a5bed3cd9ece66fb
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 25 15:46:54 2016 +0200

    Removed hosts

commit 5633cc7ca4eeb033e2848ce0cd7d6d6d19578b10
Author: Jacek Bzdak <jacek@opencraft.com>
Date:   Mon Apr 18 15:14:58 2016 +0200

    Initial version of database provisioning

Commit:
Updated all requirements to have version hashes to have repeatable builds
